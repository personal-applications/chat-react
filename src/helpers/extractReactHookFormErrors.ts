import { pascalSnakeCase } from "change-case";
import { FieldErrors, FieldValues } from "react-hook-form";

export default function extractReactHookFormErrors<T extends FieldValues>(
  errors: FieldErrors<T>,
): Record<string, string> {
  const keys = Object.keys(errors);
  const messages = Object.values(errors).map((e) => e!.message);

  const fields = keys.map((k) => {
    const pSCase = pascalSnakeCase(k);
    return pSCase.replace("_", " ");
  });

  return fields.reduce<Record<string, string>>((acc, current, index) => {
    acc[current] = messages[index] as string;
    return acc;
  }, {});
}
