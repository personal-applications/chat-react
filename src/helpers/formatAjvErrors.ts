import { FieldErrors, FieldValues } from "react-hook-form";

export function formatAjvErrors<T extends FieldValues>(
  errors: FieldErrors<T>,
): Record<keyof T, string> {
  if (!Object.keys(errors).length) {
    return {} as Record<keyof T, string>;
  }

  const result: Record<string, string> = {};
  const fields = Object.keys(errors);
  const types = Object.values(errors).map((error) => error!.type) as string[];

  types.forEach((type, index) => {
    if (type === "required") {
      result[fields[index]] = `This field is required.`;
    } else if (type === "format" && fields[index] === "email") {
      result[fields[index]] = `Invalid email address.`;
    } else if (type === "pattern") {
      result[fields[index]] = `Invalid value.`;
    } else if (type === "minLength") {
      result[fields[index]] = `Value is too short.`;
    } else if (type === "regex") {
      result[fields[index]] = `Invalid value.`;
    }
  });

  return result as Record<keyof T, string>;
}
