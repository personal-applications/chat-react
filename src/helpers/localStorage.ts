export const JWT_KEY = "jwt";

export function saveToLocalStorage<T>(key: string, value: T): void {
  localStorage.setItem(key, JSON.stringify(value));
}

export function getFromLocalStorage<T>(key: string): T | null {
  const value = localStorage.getItem(key);
  return value ? JSON.parse(value) : null;
}

export function deleteFromLocalStorage(key: string): void {
  localStorage.removeItem(key);
}