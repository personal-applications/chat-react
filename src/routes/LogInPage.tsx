import useSWRMutation from "swr/mutation";
import LogInForm, { LoginFormFields } from "../forms/LogInForm";
import { postData } from "../services/fetcher";

function LogInPage() {
  const { trigger } = useSWRMutation("/api/auth/login", postData);

  const onSubmit = async (data: LoginFormFields) => {
    await trigger(data);
  };

  return (
    <>
      <div className="h-screen w-screen bg-neutral pt-40">
        <div className="w-[600px] mx-auto">
          <h1 className="text-3xl font-semibold text-center mb-5">ChatHub</h1>
          <h1 className="text-bold text-center text-xl">Login</h1>
          <LogInForm onSubmit={onSubmit} />
        </div>
      </div>
    </>
  );
}

export default LogInPage;
