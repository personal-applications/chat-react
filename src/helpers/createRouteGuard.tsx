import React from "react";
import Redirect from "../components/Redirect";
import { JWT_KEY, getFromLocalStorage } from "./localStorage";

export type RouteGuardCondition = "loggedIn" | "loggedOut";

export default function createRouteGuard(
  C: () => React.JSX.Element,
  condition: RouteGuardCondition,
): React.JSX.Element {
  const jwt = getFromLocalStorage(JWT_KEY);
  if (!jwt && condition === "loggedIn") {
    return <Redirect to="/login" />;
  }
  if (jwt && condition === "loggedOut") {
    return <Redirect to="/home" />;
  }
  return <C />;
}
