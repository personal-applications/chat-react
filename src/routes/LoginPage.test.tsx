import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { BrowserRouter, useNavigate } from "react-router-dom";
import useSWRMutation from "swr/mutation";
import { Mock } from "vitest";
import LogInPage from "./LogInPage";
import { getFromLocalStorage } from "../helpers/localStorage";

vi.mock("swr/mutation", () => {
  return {
    default: vi.fn().mockReturnValue({}),
  };
});
vi.mock("react-router-dom", async (importOriginal) => {
  const mod = await importOriginal<typeof import("react-router-dom")>();
  return {
    ...mod,
    useNavigate: vi.fn(),
  };
});

suite("LogInPage", () => {
  test("should render the login form", () => {
    const tree = render(
      <BrowserRouter>
        <LogInPage />
      </BrowserRouter>,
    );

    expect(tree).toMatchSnapshot();
  });

  test("should navigate to /home after successful login", async () => {
    const navigateMock = vi.fn();
    (useNavigate as Mock).mockReturnValue(navigateMock);

    const triggerMock = vi.fn().mockResolvedValue({ jwt: "dummyToken" });
    (useSWRMutation as Mock).mockReturnValue({
      trigger: triggerMock,
    });

    render(
      <BrowserRouter>
        <LogInPage />
      </BrowserRouter>,
    );

    await userEvent.type(
      screen.getByPlaceholderText("Email"),
      "test@example.com",
    );
    await userEvent.type(
      screen.getByPlaceholderText("Password"),
      "Password123*",
    );
    await userEvent.click(screen.getByText("Log in"));

    expect(getFromLocalStorage("jwt")).toBe("dummyToken");
    expect(triggerMock).toHaveBeenCalledWith({
      email: "test@example.com",
      password: "Password123*",
    });
    expect(navigateMock).toHaveBeenCalledWith("/home");
  });
});
