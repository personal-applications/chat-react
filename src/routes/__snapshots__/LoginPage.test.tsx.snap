// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LogInPage > should render the login form 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="h-screen w-screen bg-neutral pt-40"
      >
        <div
          class="w-[600px] mx-auto"
        >
          <h1
            class="text-3xl font-semibold text-center mb-5"
          >
            ChatHub
          </h1>
          <h1
            class="text-bold text-center text-xl"
          >
            Login
          </h1>
          <form
            class="mt-6 flex flex-col gap-y-6 rounded-md bg-base-100 p-8"
            data-testid="login-form"
            novalidate=""
          >
            <div
              class="flex w-full flex-col"
            >
              <label
                class="input input-bordered flex items-center gap-2"
              >
                <input
                  class="grow"
                  name="email"
                  placeholder="Email"
                  type="email"
                />
              </label>
            </div>
            <div
              class="flex w-full flex-col"
            >
              <label
                class="input input-bordered flex items-center gap-2"
              >
                <input
                  class="grow"
                  name="password"
                  placeholder="Password"
                  type="password"
                />
                <button
                  type="button"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    viewBox="0 0 1024 1024"
                    width="1em"
                  >
                    <path
                      d="M396 512a112 112 0 10224 0 112 112 0 10-224 0zm546.2-25.8C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM508 688c-97.2 0-176-78.8-176-176s78.8-176 176-176 176 78.8 176 176-78.8 176-176 176z"
                    />
                  </svg>
                </button>
              </label>
            </div>
            <div
              class="ml-auto"
            >
              <a
                class="link link-primary link-hover"
                href="/forgot-password"
              >
                Forgot password?
              </a>
            </div>
            <button
              class="btn btn-primary"
              type="submit"
            >
              Log in
            </button>
          </form>
          <div
            class="flex flex-col gap-y-2 mt-3"
          >
            <span
              class="text-center"
            >
              Don't have an account?
               
              <a
                class="link link-primary link-hover"
                href="/register"
              >
                Register
              </a>
            </span>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="h-screen w-screen bg-neutral pt-40"
    >
      <div
        class="w-[600px] mx-auto"
      >
        <h1
          class="text-3xl font-semibold text-center mb-5"
        >
          ChatHub
        </h1>
        <h1
          class="text-bold text-center text-xl"
        >
          Login
        </h1>
        <form
          class="mt-6 flex flex-col gap-y-6 rounded-md bg-base-100 p-8"
          data-testid="login-form"
          novalidate=""
        >
          <div
            class="flex w-full flex-col"
          >
            <label
              class="input input-bordered flex items-center gap-2"
            >
              <input
                class="grow"
                name="email"
                placeholder="Email"
                type="email"
              />
            </label>
          </div>
          <div
            class="flex w-full flex-col"
          >
            <label
              class="input input-bordered flex items-center gap-2"
            >
              <input
                class="grow"
                name="password"
                placeholder="Password"
                type="password"
              />
              <button
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 1024 1024"
                  width="1em"
                >
                  <path
                    d="M396 512a112 112 0 10224 0 112 112 0 10-224 0zm546.2-25.8C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM508 688c-97.2 0-176-78.8-176-176s78.8-176 176-176 176 78.8 176 176-78.8 176-176 176z"
                  />
                </svg>
              </button>
            </label>
          </div>
          <div
            class="ml-auto"
          >
            <a
              class="link link-primary link-hover"
              href="/forgot-password"
            >
              Forgot password?
            </a>
          </div>
          <button
            class="btn btn-primary"
            type="submit"
          >
            Log in
          </button>
        </form>
        <div
          class="flex flex-col gap-y-2 mt-3"
        >
          <span
            class="text-center"
          >
            Don't have an account?
             
            <a
              class="link link-primary link-hover"
              href="/register"
            >
              Register
            </a>
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
